relayer/relays/execution/message/message.go

```diff
@@ -0,0 +1,320 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *Message) checkMessageVerificationResult(msgAddress common.Address, msgNonce uint64) error {
+	nonce, err := m.writer.GetLastBasicChannelNonceByAddress(msgAddress)
+	if err != nil {
+		return fmt.Errorf("fetch last basic channel message nonces by addresses: %w", err)
+	}
+
+	if nonce != msgNonce {
+		return fmt.Errorf("last basic message verification failed for address %s (expected nonce: %d, actual nonce: %d)", msgAddress, msgNonce, nonce)
+	}
+
+	log.WithFields(log.Fields{"nonce": msgNonce, "address": msgAddress}).Info("basic message verified successfully")
+	return nil
+}
+
+func (m *Message) writeIncentivizedMessages(ctx context.Context, payload ParachainPayload) error {
+	log.WithField("count", len(payload.Messages)).Info("writing incentivized messages")
+
+	for _, msg := range payload.Messages {
+		err := m.writer.WriteToParachainAndWatch(ctx, msg.Call, msg.Args...)
+		if err != nil {
+			return err
+		}
+
+		lastNonce, err := m.writer.GetLastIncentivizedChannelNonce()
+		if err != nil {
+			return err
+		}
+
+		if lastNonce != msg.Nonce {
+			return fmt.Errorf("last incentivized message verification failed (nonce: %d)", lastNonce)
+		}
+	}
+
+	return nil
+}
+
+func filterMessagesByLastNonce(messages []*chain.EthereumOutboundMessage, nonce uint64) []*chain.EthereumOutboundMessage {
+	resultMessages := []*chain.EthereumOutboundMessage{}
+
+	for _, incentivizedMessage := range messages {
+		if incentivizedMessage.Nonce <= nonce {
+			continue
+		}
+
+		resultMessages = append(resultMessages, incentivizedMessage)
+	}
+
+	return resultMessages
+}
+
+func filterMessagesByLastNonces(messages []*chain.EthereumOutboundMessage, addressNonceMap map[common.Address]uint64) []*chain.EthereumOutboundMessage {
+	resultMessages := []*chain.EthereumOutboundMessage{}
+
+	for _, basicMessage := range messages {
+		if basicMessage.Nonce <= addressNonceMap[basicMessage.Origin] {
+			continue
+		}
+
+		resultMessages = append(resultMessages, basicMessage)
+	}
+
+	return resultMessages
+}
+
+func (m *Message) syncUnprocessedBasicMessages(ctx context.Context) (uint64, error) {
+	lastVerifiedBlockNumber, err := m.writer.GetLastBasicChannelBlockNumber()
+	if err != nil {
+		return 0, fmt.Errorf("fetch last basic channel message block number: %w", err)
+	}
+
+	addressNonceMap, err := m.writer.GetLastBasicChannelNonceByAddresses(m.addresses)
+	if err != nil {
+		return 0, fmt.Errorf("fetch last basic channel message nonce: %w", err)
+	}
+
+	addressNonzeroNonceMap := make(map[common.Address]uint64, len(addressNonceMap))
+	for address, nonce := range addressNonceMap {
+		log.WithFields(log.Fields{
+			"block_number": lastVerifiedBlockNumber,
+			"address":      address,
+			"nonce":        nonce,
+		}).Info("last basic channel")
+
+		if nonce != 0 {
+			addressNonzeroNonceMap[address] = nonce
+		}
+	}
+
+	if len(addressNonzeroNonceMap) == 0 {
+		return 0, nil
```

Comment:
Similar to the incentivized channel method, if all addresses' nonces are 0, we should be able to ignore all blocks up to `lastVerifiedBlockNumber`.

```suggestion
	if len(addressNonzeroNonceMap) == 0 {
		return lastVerifiedBlockNumber, nil
```